internal/tool/tool.go

```diff
@@ -0,0 +1,183 @@
+var _ codacy.Tool = (*codacySemgrep)(nil)
+
+// Run runs the Semgrep implementation
+func (s codacySemgrep) Run(ctx context.Context, toolExecution codacy.ToolExecution) ([]codacy.Result, error) {
+	configFile, err := getConfigurationFile(*toolExecution.Patterns, toolExecution.SourceDir)
+	if err == nil {
+		defer os.Remove(configFile.Name())
+	}
+
+	// filesToAnalyse, err := getListOfFilesToAnalyse(tool.Files, sourceDir)
+	// if err != nil {
+	// 	return nil, errors.New("Error getting files to analyse: " + err.Error())
+	// }
+
+	semgrepCmd := semgrepCommand(configFile, *toolExecution.Files, toolExecution.SourceDir)
+
+	semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+	if err != nil {
+		return nil, errors.New("Error running semgrep: " + semgrepError)
+	}
+
+	result := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
+	return result, nil
+}
+
+type SemgrepOutput struct {
+	Results []SemgrepResult
+}
+
+type SemgrepResult struct {
+	CheckID       string          `json:"check_id"`
+	Path          string          `json:"path"`
+	StartLocation SemgrepLocation `json:"start"`
+	EndLocation   SemgrepLocation `json:"end"`
+	Extra         SemgrepExtra    `json:"extra"`
+}
+
+type SemgrepLocation struct {
+	Line int `json:"line"`
+}
+
+type SemgrepExtra struct {
+	Message     string `json:"message"`
+	RenderedFix string `json:"rendered_fix,omitempty"`
+}
+
+func configurationFromSourceCode(sourceFolder string) (string, error) {
+	filename := path.Join(sourceFolder, sourceConfigFileName)
+	contentByte, err := os.ReadFile(filename)
+	return string(contentByte), err
+}
+
+func writeToTempFile(content string) (*os.File, error) {
+	tmpFile, err := os.CreateTemp(os.TempDir(), "semgrep-")
+	if err != nil {
+		return nil, err
+	}
+	if _, err = tmpFile.Write([]byte(content)); err != nil {
+		return nil, err
+	}
+	if err := tmpFile.Close(); err != nil {
+		return nil, err
+	}
+
+	return tmpFile, nil
+}
+
+func getConfigurationFile(patterns []codacy.Pattern, sourceFolder string) (*os.File, error) {
+	// if no patterns, try to use configuration from source code
+	// otherwise default configuration file
+	if len(patterns) == 0 {
+		sourceConfigFileContent, err := configurationFromSourceCode(sourceFolder)
+		if err == nil {
+			return writeToTempFile(sourceConfigFileContent)
+		}
+
+		return nil, err
+	}
+
+	// TODO: generate configuration file from patterns and auto config file
+	// content := generateToolConfigurationContent(patterns)
+
+	// return writeToTempFile(content)
+	return nil, nil
+}
+
+func getConfigFileParam(configFile *os.File) []string {
+	if configFile != nil {
+		return []string{
+			"-rules",
+			configFile.Name(),
+		}
+	}
+	return []string{}
+}
+
+func commandParameters(configFile *os.File, filesToAnalyse []string) []string {
+	cmdParams := append(
+		[]string{
+			"-json", "-json_nodots",
+			"-lang", "python", // TODO: get language from toolExecution?
```

Comment:
Codacy will never invoke the tool for multiple languages at once. it shouldn't be a problem if we lookup the language from the pattern information in the tool definition

Reply:
We will need to find a way to check for the language as that definition is not present in the tool definition